Unexpected exception in solvePnPRansac caused by input points

mightbxg opened the pull request:
resolves #17799 
relates #14447

When `solvePnPRansac` is called, it will do RANSAC and then call `solvePnPGeneric`, in which `cvFindExtrinsicCameraParams2` will be called. Sometimes the point-pairs after RANSAC are less than 6, but the DLT algorithm used in `cvFindExtrinsicCameraParams2` (SOLVEPNP_ITERATIVE) needs at least 6 points, in this case the function will throw an error. See [solvePnPGeneric](https://github.com/mightbxg/opencv/blob/master/modules/calib3d/src/solvepnp.cpp#L875), [cvFindExtrinsicCameraParams2](https://github.com/mightbxg/opencv/blob/master/modules/calib3d/src/calibration.cpp#L1173).

The proper logic for `solvePnPGeneric` should be returning `false` (0 or -1) instead of throwing an error to the executable, so I added try-block, just like [SOLVEPNP_IPPE](https://github.com/mightbxg/opencv/blob/941a9792f7c7d60640c82926091d710e6ecbfb55/modules/calib3d/src/solvepnp.cpp#L891) below.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [ ] There is reference to original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

vpisarev commented:
@mightbxg, thank you for the contribution. Maybe, inserting try-catch into `solvePnPGeneric` is not a good idea. I suggest to insert try-catch into `solvePnPRansac` or some RANSAC callbacks if it uses ones (I can't remember it right now). This way you will make a pure bugfix instead of bugfix+compatibility break.

BTW, please, keep in mind that in the upcoming OpenCV 5.0 (the "next" branch) the implementation of solvePnP, including the RANSAC flavor, will change. And both in the master branch and the next branch there is new RANSAC framework (called USAC) available that provides much better performance. You are welcome to try it out. If it works well on your dataset, then switching to this USAC is probably a better idea than trying to fix the obsolete solution that will eventually be thrown away.

vpisarev reviewed (changes requested):
need to move try-catch outside of solvePnPGeneric into solvePnPRansac. Otherwise some user may experience undefined behaviour instead of exception in the corner cases

asmorkalov commented:
@mightbxg Friendly reminder.

mightbxg commented:
@vpisarev Thanks a lot for you suggestions. I think you are right about not modifying `solvePnPGeneric`, so I removed the try-catch block and added a point-num-checking in `solvePnPRansac` before it calls `solvePnP`. The commit is now amended.

[Here](https://github.com/opencv/opencv/files/5872501/pts.zip) is the test data that can reproduce the problem. Load points from the yaml file and call `solvePnPRansac` using params below, then an error will be thrown out (OpenCV 3.4.12 and 4.5.1 are both affected):
```cpp
Mat rvec, tvec;
bool ret = solvePnPRansac(pts3d, pts2d, Mat::eye(3, 3, CV_64FC1), noArray(), rvec, tvec, false, 100, 4. / 460.0);
```

I looked into the document about USAC and it's quite exciting, I will try it later. However, as the traditional `solvePnPRansac` in OpenCV 3&4 is widely used and may be used for much longer time, I think it's necessary to fix this issue.

mightbxg reviewed (commented):
Removed try-catch block from solvePnPGeneric. Added a point-num-checking in solvePnPRansac before it calls solvePnP. The commit is amended.

vpisarev commented:
@mightbxg, thank you!
Since you are on it, may I ask you to add a tiny test to
https://github.com/opencv/opencv/blob/3.4/modules/calib3d/test/test_solvepnp_ransac.cpp
that does what you've described?

1. initialize 2d point and 3d point arrays, e.g. 5 points in each
2. then calls solvePnPRansac enclosed in `EXPECT_THROW`:

```
EXPECT_THROW(solvePnPRansac(...), cv::Exception);
```

mightbxg commented:
> @mightbxg, thank you!
> Since you are on it, may I ask you to add a tiny test to
> https://github.com/opencv/opencv/blob/3.4/modules/calib3d/test/test_solvepnp_ransac.cpp
> that does what you've described?
> 
> 1. initialize 2d point and 3d point arrays, e.g. 5 points in each
> 2. then call solvePnPRansac enclosed in `EXPECT_THROW`:
> 
> ```
> EXPECT_THROW(solvePnPRansac(...), cv::Exception);
> ```

@vpisarev Sorry, I'm not sure about the purpose of adding this test. Is it for showing that the problem was there before fix (as I described before) or the problem is gone after fix? 

To clarify, let me describe the problem again. The role of "Exception" is to inform the user of his/her "bad usage". For instance, P3P need 4 points, and when 3 points are input, there should be an Exception to tell the user that the usage is wrong.
However, an Exception caused by intermediate calculation is unexpected. When I call `solvePnPRansac` with 69 points ([data I provided above](https://github.com/opencv/opencv/files/5872501/pts.zip)), then I should get a result (even if it's `false`) rather than an Exception. Because I cannot determine if the input data is "wrong" before RANSAC calculation.

For the former purpose (reproduce the problem), should I just create an "issue" based on the not-fixed branch 3.4?

For the later purpose, I presume this test should be added to the pull request. But as the problem is already fixed, `solvePnPRansac` will return `false` instead of throwing `cv::Exception`. Maybe I should enclose solvePnPRansac calling with `EXPECT_FALSE` instead of `EXPECT_THROW`? Moreover, should I add a new commit for the test or squash it to the existing commit?

catree commented on modules/calib3d/src/solvepnp.cpp:
```diff
@@ -309,6 +309,17 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
     int npoints1 = compressElems(&opoints_inliers[0], mask, 1, npoints);
     compressElems(&ipoints_inliers[0], mask, 1, npoints);
 
+    if ( flags == SOLVEPNP_ITERATIVE && npoints1 < 6 )
```
I believe it is a little bit more complex.

When 3D points are not planar, `cvFindExtrinsicCameraParams2` uses DLT algorithm which indeed requires at least 6 points.

But with coplanar 3D points, `cvFindExtrinsicCameraParams2` should work with at least 4 points.

---

BTW, see here: https://github.com/opencv/opencv/issues/8782 my previous comment about the behavior of `solvePnPRansac` and this PR: https://github.com/opencv/opencv/pull/9086

In my opinion, behavior of `solvePnPRansac` is still not totally clear / intuitive.

mightbxg replied:
Ok, then `cvFindExtrinsicCameraParams2` may get good result with 4 points, but may throw Exception with 5 points.

I think CV_CheckGE in [`cvFindExtrinsicCameraParams2`](https://github.com/opencv/opencv/blob/863ecded30dcaedff0ece893277411f39bb3c1c6/modules/calib3d/src/calibration.cpp#L1099) is not a good idea. But since it returns void (not bool), throwing Exception may be the only option to inform that the result is bad.

As a result, try-catch seems inevitable.

catree replied:
> Ok, then `cvFindExtrinsicCameraParams2` may get good result with 4 points, but may throw Exception with 5 points.

Not sure to follow correctly. In `cvFindExtrinsicCameraParams2` it checks first if the 3D points are coplanar. For coplanar case, it should never throw exception, except theoretically for degenerate cases like all collinear points but this is not handled/detected.

And there are already a check for at least 4 points if `useExtrinsicGuess` is false: https://github.com/opencv/opencv/blob/863ecded30dcaedff0ece893277411f39bb3c1c6/modules/calib3d/src/calibration.cpp#L1014

mightbxg replied:
`cvFindExtrinsicCameraParams2` does check if the 3D points are coplanar. And if not, it may throw exception.

When some one calls `cvFindExtrinsicCameraParams2` directly, he can make sure the input points are more than 4. If he knows that the points may not be coplanar, he can make sure the points are more than 6. In this cases, `cvFindExtrinsicCameraParams2` won't throw exceptions, because the user calls it in the "right" way.

However, codes in `solvePnPRansac` does not guarantee that. When the user knows the points used to solve PnP are very likely to be noncoplanar, the only thing he can do is to input as many points as possible (6 at least). But RANSACE procedure (the intermediate calculation) may reduce the points and make them not suitable for `cvFindExtrinsicCameraParams2`. That's why I think Exception here is not supposed to happen. It's NOT the user's fault.

To fix this, `solvePnPRansac` should either check the coplanarity before calling `solvePnP`, or call `solvePnP` with try-catch. The former solution may add unnecessary calculation.

alalek commented:
> purpose of adding this test.

Main purpose is avoiding same/similar regressions in the future.

> that the problem was there before fix
> the problem is gone after fix

These cases should work too (as described - fail before, pass after the fix).

mightbxg commented:
I mistakenly force-pushed branch 3.4 to my own branch (mightbxg:bugfix_PnPRansac), and github closed this pull request automatically. Sorry for that. I will reopen this later.

mightbxg commented:
@catree Calling `solvePnP` with try-catch in `solvePnPRansac` may be better. Test case is added. @alalek @vpisarev

catree commented:
@mightbxg 

> Calling `solvePnP` with try-catch in `solvePnPRansac` may be better.

I have not re-read the code carefully, but one possible other option could be to log somehow the error to warn the user that `SOLVEPNP_ITERATIVE` was not successful but still returns `rvec` and `tvec` solution computed from the `Minimal Sample Sets` (MSS)?

Because if I recall correctly, to compute solution for the MSS, 5 points are used and EPnP method. Consensus set is also 5 points, so it should be possible to return directly `rvec` and `tvec` computed during the MSS step.

mightbxg commented:
@catree You are right about the MSS step. It can calculate and return `rvec` and `tvec`, even if error occurs in the following steps.

The user can know if `SOLVEPNP_ITERATIVE` was successful or not by checking the returned `bool` value of `solvePnPRansac`. 

As a computing library, I don't think OpenCV should "log something" for the user.

catree commented:
@mightbxg 

> The user can know if `SOLVEPNP_ITERATIVE` was successful or not by checking the returned `bool` value of `solvePnPRansac`.

The thing is that if `solvePnPRansac` returns `false`, you are not supposed to use `rvec` and `tvec` passed in parameters.
How could you differentiate between the method completely fails and this very specific case where the consensus set is 5 points correctly estimated, but `SOLVEPNP_ITERATIVE` later fails?

(I don't have the solution.)

(`try/catch` approach seems fine for me at first sight.)

mightbxg commented:
@catree 
> How could you differentiate between the method completely fails and this very specific case where the consensus set is 5 points correctly estimated, but `SOLVEPNP_ITERATIVE` later fails?

I got your point. Your opinion is that if the MSS stage passed (EPNP for 5 or more points), then `solvePnPRansac` should return `true`, even if `SOLVEPNP_ITERATIVE` stage fails. Because the result of MSS stage is still "correct" to use.

I agree with that in some degree. But since the user calls `solvePnPRansac` with `flags = SOLVEPNP_ITERATIVE`(not `SOLVEPNP_EPNP`) and input points much more than 6, maybe he only wants the result from `ITERATIVE` stage (EPNP may have lower precision)?

So, this question seems a little subjective and may need more discussion. [The existing code](https://github.com/opencv/opencv/blob/0be18f5cb0aa6e45febde51ea7c8af253d03f144/modules/calib3d/src/solvepnp.cpp#L318-L327) in `solvePnPRansac` prefer to return `false` if `SOLVEPNP_ITERATIVE` stage fails, and my pull request didn't change that.

catree commented:
@mightbxg

I see two approaches with this specific case:
- log using the [OpenCV logging](https://docs.opencv.org/4.5.1/da/db0/namespacecv_1_1utils_1_1logging.html) system an info/debug/verbose message to state that `SOLVEPNP_ITERATIVE` failed due to non coplanar 3D data and less than 6 points, but result from `EPnP` is still returned
- return `EPnP` results and return `true`

or:
- return `false`
- and also add log info (it does not hurt to add logging info, for instance `LOG_LEVEL_DEBUG` level will only print in `debug` build mode)

---

I have a slight preference for the first approach, but second solution is fine also for me.

In any case, root case of this specific situation should be investigated, because:
- if you have a bunch of 3D data
- but the RANSAC method failed to find a consensus set bigger than 5 points
- it means that the root case can come from the 3D data, 2D data (e.g. very bad 2D features matching, or bad RANSAC parameters (e.g. too strict reprojection error threshold?)

mightbxg commented:
@catree
All right, I think you are right, and I took your first suggestion, commit is now amended.

With this change, the user can choose to accept the result when `solvePnPRansac` returns true, even if it's calculated by MSS stage. He can also choose to reject it if he thinks the result is not good enough by checking the number of inliers (Apparently, 5 inliers means the result cannot be from `ITERATIVE` stage).

BTW, I noticed someone else is working on this issue too #19492. His test data seems more concise.

mightbxg reviewed (commented):
`solvePnPRansac` returns true even if the `solvePnP` stage failed. In this case, result from MSS stage is taken, and the exception content from `solvePnP` is put into log.

catree commented on modules/calib3d/src/solvepnp.cpp:
```diff
@@ -311,19 +311,23 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
+    catch (const cv::Exception& e)
+    {
+        result = 0;
+        CV_UNUSED(e);
```
Is it needed since it is used in `CV_LOG_DEBUG`?

mightbxg replied:
I'm not quite sure about this, but I have seen same usage in [calibinit.cpp](https://github.com/opencv/opencv/blob/cac5b0fe5a3e06718a6cb408d6468ff87428ef2a/modules/calib3d/src/calibinit.cpp#L2261-L2266). Maybe in Release build, `CV_LOG_DEBUG` doesn't use `e`, then `CV_UNUSED` can supress the warning?

catree replied:
Yes, I think that it is to probably suppress warning in release mode.

catree commented on modules/calib3d/test/test_solvepnp_ransac.cpp:
```diff
@@ -837,6 +837,92 @@ TEST(Calib3d_SolvePnPRansac, double_support)
+    Mat camera_mat = Mat::eye(3, 3, CV_64FC1);
+    Mat rvec, tvec;
+    vector<uchar> inliers;
+    bool result = solvePnPRansac(pts3d, pts2d, Mat::eye(3, 3, CV_64FC1), noArray(), rvec, tvec, false, 100, 4. / 460.0, 0.99, inliers);
```
> 4. / 460.0

Reprojection error is in pixel unit (default value is 8). Wondering if it explains why you only have a consensus set of 5 points?

mightbxg replied:
In my test data, the 2D points are from the normalized image coordiante system, that's why the camera matrix is `Mat::eye(3, 3, CV_64FC1)`. So, the reprojection error here is not in pixel unit (`1.0/f` pixel in fact).

As you can see, the focal length of my camera is about 460.0, and the tolerance of reprojection error is set to 4 pixels. It should be fine.

catree replied:
You are right. I did not see that the camera matrix is identity.

catree reviewed (commented):
Quick review.

Yes, probably we could try to fuse the two PRs.

/cc @Johnzdh

catree commented on modules/calib3d/test/test_solvepnp_ransac.cpp:
```diff
@@ -837,6 +837,92 @@ TEST(Calib3d_SolvePnPRansac, double_support)
+    vector<Point3f> pts3d;
+    vector<Point2f> pts2d;
+
+    pts3d = {
```
See the buildbot (https://pullrequest.opencv.org/#/summary/opencv) errors.

This requires C++11 features.

mightbxg replied:
Fixed now.

mightbxg reviewed (commented):
Changes:
- Store test data with cv::Mat instead of std::vector to avoid c++11 feature (initializer list).
- Reduce test data to minimum set (6 input points)

catree commented on modules/calib3d/src/solvepnp.cpp:
```diff
@@ -311,19 +311,23 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
+    {
+        result = 0;
+        CV_UNUSED(e);
+        CV_LOG_DEBUG(NULL, "solvePnPRansac(): solvePnP stage failed, use result from MSS(Minimal Sample Sets) stage instead:\n" << e.what());
```
Maybe something like this? This is to explain to the user that everything went fine (consensus set was found that agree with the hypothesis), but issue occurs when computing the final pose using all the points in the consensus set.

```suggestion
        CV_LOG_DEBUG(NULL, "solvePnPRansac(): solvePnP stage to compute the final pose using points in the consensus set failed, use result from MSS (Minimal Sample Sets) stage instead:\n" << e.what());
```

catree commented on modules/calib3d/test/test_solvepnp_ransac.cpp:
```diff
@@ -837,6 +837,34 @@ TEST(Calib3d_SolvePnPRansac, double_support)
+
+TEST(Calib3d_SolvePnPRansac, bad_input_points)
+{
+    Mat pts2d = (Mat_<float>(6, 2) <<
```
I would add some comments before this line to explain this test:

- with this specific data
- when computing the final pose using points in the consensus set with `SOLVEPNP_ITERATIVE` and `solvePnP()`
- an exception is thrown from `solvePnP` because there are 5 non-coplanar 3D points and the DLT algorithm needs at least 6 non-coplanar 3D points
- with PR #19253 we choose to return `true`, with the pose estimated from the MSS stage instead of throwing the exception

mightbxg replied:
Thanks, it do need some explanation.

catree reviewed (commented):
Just some minor suggestions.

Thanks for your contribution. I am also an external contributor and it is great to see contributions from the community.

mightbxg commented:
@catree @vpisarev @asmorkalov 
Please let me know if further modification is needed.

vpisarev commented:
thank you, looks good to me! :+1:

alalek commented on modules/calib3d/src/solvepnp.cpp:
```diff
@@ -311,19 +311,24 @@ bool solvePnPRansac(InputArray _opoints, InputArray _ipoints,
-        if( _inliers.needed() )
-            _inliers.release();
-
-        return false;
```
This part changes behavior of the case which is not covered by PR's description:
- `result = solvePnP()` doesn't throw exception
- it returns `result <= 0`

Alternative #19492 preserves behavior of non-exception case.

---

> When I call solvePnPRansac with 69 points (data I provided above), then I should get a result (even if it's false) rather than an Exception.

`false` code path is changed.

> solvePnPRansac returns true even if the solvePnP stage failed. In this case, result from MSS stage is taken, and the exception content from solvePnP is put into log.

Only exception's case is logged, but returned value is forced "true" for other cases too.

mightbxg replied:
You are right, the logging strategy here is problematic.

alalek replied:
Main question is about the final returned value.

|result = solvePnP|existed code|This PR| PR 19492|case4(no PR)
|---|---|---|---|---|
|result > 0|true|true|true|true|
|result<=0|false|true (&#x2753;) |false|false|
|exception|exception|true (&#x2753;)|true (&#x2753;)| false|

asmaloney replied:
I think you meant to @ someone else (not me)?

alalek reviewed (changes requested):
@vpisarev @asmaloney 
Need to discuss the change below.

Function result values:

|result = solvePnP|(1) existed code|(2) This PR| (3) PR 19492|case4(no PR)
|---|---|---|---|---|
|result > 0|true|true|true|true|
|result<=0|false|true (&#x2753;) |false|false|
|exception|exception|true (&#x2753;)|true (&#x2753;)| false|

mightbxg commented:
@alalek 
I have changed the logging strategy so that non-exception case can also be logged.

In conclusion, this PR introduced changes about `solvePnPRansac`:
1. The case that error occurs (an Exception is thrown) in solvePnP stage is treated as `solvePnP()` failing (no solution returned by `solvePnPGeneric`), and the error info is logged.
2. `solvePnPRansac` is considered as `success` (it returns true) when MSS stage passed but solvePnP stage failed (no solution or error occurs). In this case, result from MSS stage is taken, and an warning message is logged.

The second change is what should be discussed here. Like you've said, the behavior of non-exception case in the existing OpenCV codes is to return `false`. But I personally agree with @catree, that it should return `true` to tell the user that the result from MSS stage is also "correct" to use.

alalek commented:
So, what is the purpose of this `solvePnP()` call? Does it becomes as optional result refinement?
Should be guarantee that result is feasible after the main RANSAC code?

Why we can't just tune `model_points` ( = 5 by default for now) ?

mightbxg commented:
@alalek 
1. With this PR's changes, the solvePnP stage indeed becomes optional result-refinement. However, user can know if the result is from solvePnP stage or MSS stage by checking the number of inliers.
2. In the [existing codes](https://github.com/opencv/opencv/blob/8f6ba5a089a61d62bc2d06a8360960453f448a35/modules/calib3d/src/solvepnp.cpp#L248-L274), when `solvePnPRansac` is called with exactly 5 input points and `flags=SOLVEPNP_ITERATIVE`, it will call `solvePnP` with `flags=SOLVEPNP_EPNP` (just like the MSS stage), and return `ture` (with EPnP result). So, I think the result after RANSAC is just fine.
3. Increasing `model_points` is a potential solution (need discussion). But if we change it to `6`, then what will we do when the number of input points is 5? Should we use other ransac kernel method rather than EPnP or just return false?

catree commented:
Just did a very quick look on the new comments. Need to re-read everything later in details, and think more about the behavior changes / matrix.

---

> Increasing model_points is a potential solution (need discussion).

Normally, the RANSAC algorithm is based upon the assumption to use the minimal number of data to compute a solution, and then check if this solution agrees with the whole data. The solution that gives the bigger consensus set is retained.
Then, the final solution is computed using all the data in the consensus set.

So increasing the number of `model_points` to `6` will increase the combinatorial. From my knowledge, `4` points is enough for 6D pose computation. In OpenCV, `5` points is used because I think it is recommended to use `5` points in the `EPnP` paper, and `EPnP` is the default method for the MSS stage (there are exceptions depending on the input `solvePnPRansac` parameters...).

There are I think some workshop on this topic:

- [Minimal Problems in Computer Vision](http://aag.ciirc.cvut.cz/minimal/)
- [The Art of Solving Minimal Problems](http://cmp.felk.cvut.cz/minimal-cvpr-2019/)

catree commented:
We should try to fix only this specific case:

- somehow the biggest consensus set is only 5 points
- these five 3D points are not coplanar, thus calling `solvePnP()` with `SOLVEPNP_ITERATIVE` flag will throw an exception

In my opinion, computing the final pose with `SOLVEPNP_ITERATIVE` flag should use `useExtrinsicGuess=true` and `rvec`, `tvec` estimated during the MSS step. Because this consensus set has been found using 3D points with reprojection errors below a threshold, so no need to compute again an initial solution before the refinement. This was one of my interrogations in #8782

---

So in this specific case, since it is not possible mathematically to compute 3D pose with 5 points and the DLT algorithm, I propose to:

- somehow detect this case: by catching the exception and comparing the exception message?, by checking if the 3D points are coplanar?
- and instead of returning `rvec`, `tvec` from the MSS step
- use `solvePnP()` with `SOLVEPNP_ITERATIVE` flag and `useExtrinsicGuess=true`

And not change the behavior for the other cases:
- e.g. if the biggest consensus set is greater than 5 points but an exception is thrown, the same behavior before / after this PR should be retain
- maybe this would allow detecting these new failure cases from users, and fix the algorithm if possible

mightbxg commented:
The initial reason of this PR is: I found my SLAM program crashes frequently because of the 5-points-issue, and it cannot be avoided outside OpenCV without try-catch, which I think is unreasonable.

The main problem for me is the unexpected exception, not the returned value in this case. I (and maybe most of SLAM developers who use `solvePnPRansac`) don't care about the returned value if inliers are as few as 5. In fact, I reject the PnP results with inliers fewer than 10, because they may add more measurement errors to the SLAM system.

On the other hand, as @vpisarev have said, `solvePnPRansac` may eventually be replaced by USAC. Is it necessary to modify `solvePnPRansac` such a lot?

kundanSingh11 commented:
still  getting this issue in android 

OpenCV(4.10.0) Error: Unspecified error (> DLT algorithm needs at least 6 points for pose estimation from 3D-2D point correspondences. (expected: 'count >= 6'), where
                                                                                                    >     'count' is 4
                                                                                                    > must be greater than or equal to
                                                                                                    >     '6' is 6
                                                                                                    ) in void cvFindExtrinsicCameraParams2(const CvMat *, const CvMat *, const CvMat *, const CvMat *, CvMat *, CvMat *, int), file /home/ci/opencv/modules/calib3d/src/calibration.cpp, line 1280


Any help ?
